core/src/pollers/poll_buffer.rs

```diff
@@ -186,12 +379,218 @@ where
+    if let Err(e) = t.await {
+        if e.is_panic() {
+            let as_panic = e.into_panic().downcast::<String>();
+            dbg_panic!(
+                "Poller task died or did not terminate cleanly: {:?}",
+                as_panic
+            );
+        }
+    }
+}
+
+struct ActiveCounter<F: Fn(usize)>(watch::Sender<usize>, Option<Arc<F>>);
+impl<F> ActiveCounter<F>
+where
+    F: Fn(usize),
+{
+    fn new(a: watch::Sender<usize>, change_fn: Option<Arc<F>>) -> Self {
+        a.send_modify(|v| {
+            *v += 1;
+            if let Some(cfn) = change_fn.as_ref() {
+                cfn(*v);
+            }
+        });
+        Self(a, change_fn)
+    }
+}
+impl<F> Drop for ActiveCounter<F>
+where
+    F: Fn(usize),
+{
+    fn drop(&mut self) {
+        self.0.send_modify(|v| {
+            *v -= 1;
+            if let Some(cfn) = self.1.as_ref() {
+                cfn(*v)
+            };
+        });
+    }
+}
+
+struct PollScaler<F> {
+    report_handle: Arc<PollScalerReportHandle>,
+    active_tx: watch::Sender<usize>,
+    active_rx: watch::Receiver<usize>,
+    num_pollers_handler: Option<Arc<F>>,
+    ingestor_task: Option<JoinHandle<()>>,
+}
+
+impl<F> PollScaler<F>
+where
+    F: Fn(usize),
+{
+    fn new(
+        behavior: PollerBehavior,
+        num_pollers_handler: Option<F>,
+        shutdown: CancellationToken,
+    ) -> Self {
+        let (active_tx, active_rx) = watch::channel(0);
+        let num_pollers_handler = num_pollers_handler.map(Arc::new);
+        let (min, max, target) = match behavior {
+            PollerBehavior::SimpleMaximum(m) => (1, m, m),
+            PollerBehavior::Autoscaling {
+                minimum,
+                maximum,
+                initial,
+            } => (minimum, maximum, initial),
+        };
+        let report_handle = Arc::new(PollScalerReportHandle {
+            max,
+            min,
+            target: AtomicUsize::new(target),
+            ever_saw_scaling_decision: AtomicBool::default(),
+            behavior,
+            ingested_this_period: Default::default(),
+            ingested_last_period: Default::default(),
+            scale_up_allowed: AtomicBool::new(true),
+        });
+        let rhc = report_handle.clone();
+        let ingestor_task = if behavior.is_autoscaling() {
+            Some(tokio::task::spawn(async move {
+                let mut interval = tokio::time::interval(Duration::from_millis(100));
+                loop {
+                    tokio::select! {
+                        _ = interval.tick() => {}
+                        _ = shutdown.cancelled() => { break; }
+                    }
+                    let ingested = rhc.ingested_this_period.swap(0, Ordering::Relaxed);
+                    let ingested_last = rhc.ingested_last_period.swap(ingested, Ordering::Relaxed);
+                    rhc.scale_up_allowed
+                        .store(ingested_last >= ingested, Ordering::Relaxed);
+                }
+            }))
```

Comment:
So to confirm, the algorithm is:

Server returns amount of pollers to scale up (or down if negative) on each poll response, and SDK respects that decision so long as it's bounded by min/max, and in the case of scale up, there were at least as many polls accepted this 100ms period as last 100ms?

Reply:
I'm glad you looked at this again because it sort of obviously makes no sense. I was wondering why my testing wasn't showing good restriction of overshooting like I know it did at one point, and I had tried like a bazillion different methods but I knew this worked and it was simple, so I went back to it -- but results were inconsistent.

Somehow this comparison got flipped at some point. I've changed it back and the results are much better again.

Reply:
Honestly though I'm tempted to get rid of this entirely. My tests show it can still be "defeated" semi often by rapid scale ups, and ingestion is indeed going up so there's not really any reason to say no, and then you end up with a bunch of polls sitting there once the backlog clears.

Smoothing out the calculation so that the shorter timeout gets set more reliably might have more value

Reply:
Agreed. I am always a bit wary when I see hardcoded time expectations like 100ms unless that's a good number arrived by testing or something (which maybe it is). Maybe a more naive "do not make poller count changes more frequently than X interval" can help, but I don't understand the details and haven't run the tests.